File: go/borsh.go
// Borsh v1 encode/decode, byte-identical to Rust `borsh::to_vec`. This is the
// wire format the engine speaks for calldata, return values, storage values,
// and event data — so every byte here must match Rust exactly (proven by the
// golden-vector parity test against the shared vectors/codec_borsh.json).
//
// Rules (HOST_FN_ABI_SPEC §14): integers are fixed-width little-endian,
// two's-complement for signed; bool is 1 byte (0/1); address/hash32/[u8;N] are
// N raw bytes with NO length prefix; bytes and string are u32-LE length prefix
// + payload (string length is UTF-8 byte count); vec<T> is u32-LE count +
// elements back-to-back; structs/tuples are frameless (fields concatenated in
// declaration order, no wrapper). Pure Go — natively testable, TinyGo-clean.

package pyde

// Encoder builds a borsh byte stream. Field methods return the encoder for
// chaining; call Finish to read the result.
type Encoder struct{ buf []byte }

// NewEncoder returns a fresh Encoder.
func NewEncoder() *Encoder { return &Encoder{} }

// Finish returns the accumulated borsh bytes.
func (e *Encoder) Finish() []byte { return e.buf }

// Len reports how many bytes have been written so far.
func (e *Encoder) Len() int { return len(e.buf) }

func (e *Encoder) U8(v uint8) *Encoder { e.buf = append(e.buf, v); return e }

func (e *Encoder) U16(v uint16) *Encoder {
	e.buf = append(e.buf, byte(v), byte(v>>8))
	return e
}

func (e *Encoder) U32(v uint32) *Encoder {
	e.buf = append(e.buf, byte(v), byte(v>>8), byte(v>>16), byte(v>>24))
	return e
}

func (e *Encoder) U64(v uint64) *Encoder {
	e.buf = append(e.buf,
		byte(v), byte(v>>8), byte(v>>16), byte(v>>24),
		byte(v>>32), byte(v>>40), byte(v>>48), byte(v>>56))
	return e
}

func (e *Encoder) U128(v U128) *Encoder {
	b := v.ToBytesLE()
	e.buf = append(e.buf, b[:]...)
	return e
}

// Signed integers are two's-complement — identical bytes to the unsigned bit
// pattern, so we reinterpret and reuse the unsigned writers.
func (e *Encoder) I8(v int8) *Encoder   { return e.U8(uint8(v)) }
func (e *Encoder) I16(v int16) *Encoder { return e.U16(uint16(v)) }
func (e *Encoder) I32(v int32) *Encoder { return e.U32(uint32(v)) }
func (e *Encoder) I64(v int64) *Encoder { return e.U64(uint64(v)) }
func (e *Encoder) I128(v I128) *Encoder {
	b := v.ToBytesLE()
	e.buf = append(e.buf, b[:]...)
	return e
}

func (e *Encoder) Bool(v bool) *Encoder {
	if v {
		e.buf = append(e.buf, 1)
	} else {
		e.buf = append(e.buf, 0)
	}
	return e
}

// Address / Bytes32: 32 raw bytes, no length prefix.
func (e *Encoder) Address(a Address) *Encoder { e.buf = append(e.buf, a[:]...); return e }
func (e *Encoder) Bytes32(h Bytes32) *Encoder { e.buf = append(e.buf, h[:]...); return e }

// Bytes: u32-LE length prefix + raw payload (borsh Vec<u8>).
func (e *Encoder) Bytes(b []byte) *Encoder {
	e.U32(uint32(len(b)))
	e.buf = append(e.buf, b...)
	return e
}

// String: u32-LE UTF-8-byte-length prefix + UTF-8 payload.
func (e *Encoder) String(s string) *Encoder {
	e.U32(uint32(len(s)))
	e.buf = append(e.buf, s...)
	return e
}

// VecU64: u32-LE count + each element as 8-byte LE.
func (e *Encoder) VecU64(vs []uint64) *Encoder {
	e.U32(uint32(len(vs)))
	for _, v := range vs {
		e.U64(v)
	}
	return e
}

// Decoder reads a borsh byte stream with a forward cursor. Every read is
// bounds-checked; an underrun panics, which on-chain routes to `pyde.revert`
// (so malformed/short calldata reverts rather than reading out of bounds).
type Decoder struct {
	buf []byte
	pos int
}

// NewDecoder returns a Decoder over b.
func NewDecoder(b []byte) *Decoder { return &Decoder{buf: b} }

// Remaining reports how many bytes are left unread.
func (d *Decoder) Remaining() int { return len(d.buf) - d.pos }

func (d *Decoder) take(n int) []byte {
	if n < 0 || d.pos+n > len(d.buf) {
		panic("pyde: borsh decode underrun")
	}
	s := d.buf[d.pos : d.pos+n]
	d.pos += n
	return s
}

func (d *Decoder) U8() uint8   { return d.take(1)[0] }
func (d *Decoder) U16() uint16 { return u16LE(d.take(2)) }
func (d *Decoder) U32() uint32 { return u32LE(d.take(4)) }
func (d *Decoder) U64() uint64 { return u64LE(d.take(8)) }
func (d *Decoder) U128() U128  { return U128FromBytesLE(d.take(16)) }
func (d *Decoder) I8() int8    { return int8(d.take(1)[0]) }
func (d *Decoder) I16() int16  { return int16(u16LE(d.take(2))) }
func (d *Decoder) I32() int32  { return int32(u32LE(d.take(4))) }
func (d *Decoder) I64() int64  { return int64(u64LE(d.take(8))) }
func (d *Decoder) I128() I128  { return I128FromBytesLE(d.take(16)) }

// Bool decodes a strict borsh bool: only 0x00/0x01 are valid (matching Rust,
// which rejects other bytes).
func (d *Decoder) Bool() bool {
	b := d.take(1)[0]
	if b > 1 {
		panic("pyde: borsh invalid bool")
	}
	return b == 1
}

func (d *Decoder) Address() Address { return AddressFromBytes(d.take(AddressLen)) }
func (d *Decoder) Bytes32() Bytes32 { return Bytes32FromBytes(d.take(Bytes32Len)) }

// Bytes decodes a length-prefixed byte string. The result aliases the decoder
// buffer (read-only calldata); copy if you need to retain it past the frame.
func (d *Decoder) Bytes() []byte { return d.take(int(d.U32())) }

// String decodes a length-prefixed UTF-8 string (the bytes are copied).
func (d *Decoder) String() string { return string(d.take(int(d.U32()))) }

// VecU64 decodes a count-prefixed slice of u64.
func (d *Decoder) VecU64() []uint64 {
	n := int(d.U32())
	out := make([]uint64, n)
	for i := range out {
		out[i] = d.U64()
	}
	return out
}

File: assemblyscript/assembly/borsh.ts
// Borsh v1 encode / decode for AssemblyScript — byte-identical to the
// Rust reference's `borsh::to_vec` (rust/pyde-host + the `borsh` crate).
//
// ─── The rules (borsh v1, the subset Pyde's ScalarType vocab uses) ────
//
//   uN / iN          fixed-width little-endian (u8..u128, i8..i128)
//   bool             1 byte, 0x00 = false, 0x01 = true
//   address / hash32 32 RAW bytes, NO length prefix (a fixed `[u8; 32]`)
//   bytes            u32_le(len) ‖ raw bytes
//   string           u32_le(byte_len) ‖ UTF-8 bytes  (no NUL)
//   vec<T>           u32_le(len) ‖ element[0] ‖ element[1] ‖ ...
//   tuple / struct   field[0] ‖ field[1] ‖ ...   (NO framing)
//
// The last rule is why there is no `tuple()` method: encoding a struct is
// just calling one field method per field, in declaration order. That
// frameless concatenation is exactly `#[pyde::entry]`'s calldata layout,
// so an encoder built field-by-field produces calldata a Rust contract
// decodes verbatim.
//
// This layer OWNS nothing about byte order itself — it composes the
// slice-1 `codec` (LE ints) and slice-2 `u128` codec. Fix a byte there,
// fix it everywhere.

import {
  writeU8LE, writeU16LE, writeU32LE, writeU64LE,
  writeI8LE, writeI16LE, writeI32LE, writeI64LE,
  readU8LE, readU16LE, readU32LE, readU64LE,
  readI8LE, readI16LE, readI32LE, readI64LE,
} from "./codec";
import {
  u128, i128,
  writeU128LE, writeI128LE, readU128LE, readI128LE,
} from "./u128";
import { Address, Bytes32, bytes32FromPtr, BYTES32_LEN } from "./types";

// ─────────────────────────────────────────────────────────────────────
// Encoder
// ─────────────────────────────────────────────────────────────────────
//
// Backed by an ArrayBuffer we grow ourselves (doubling). `reserve(n)`
// grows if needed, then returns a pointer to the freshly-claimed tail —
// valid until the NEXT reserve. Every writer calls reserve immediately
// before writing, so the pointer is never stale. `this.data` is an
// ArrayBuffer reference (a GC root), so the backing bytes stay alive.

export class BorshEncoder {
  private data: ArrayBuffer;
  private len: i32;

  constructor(initialCapacity: i32 = 64) {
    this.data = new ArrayBuffer(initialCapacity > 0 ? initialCapacity : 1);
    this.len = 0;
  }

  // Grow to fit `n` more bytes, return a pointer to the tail, advance len.
  private reserve(n: i32): usize {
    const need = this.len + n;
    const cap = this.data.byteLength;
    if (need > cap) {
      let newCap = cap;
      while (newCap < need) newCap <<= 1;
      const grown = new ArrayBuffer(newCap);
      memory.copy(changetype<usize>(grown), changetype<usize>(this.data), this.len);
      this.data = grown;
    }
    const p = changetype<usize>(this.data) + <usize>this.len;
    this.len += n;
    return p;
  }

  // ── fixed-width integers ────────────────────────────────────────────
  u8(v: u8): BorshEncoder { writeU8LE(this.reserve(1), v); return this; }
  u16(v: u16): BorshEncoder { writeU16LE(this.reserve(2), v); return this; }
  u32(v: u32): BorshEncoder { writeU32LE(this.reserve(4), v); return this; }
  u64(v: u64): BorshEncoder { writeU64LE(this.reserve(8), v); return this; }
  u128(v: u128): BorshEncoder { writeU128LE(this.reserve(16), v); return this; }

  i8(v: i8): BorshEncoder { writeI8LE(this.reserve(1), v); return this; }
  i16(v: i16): BorshEncoder { writeI16LE(this.reserve(2), v); return this; }
  i32(v: i32): BorshEncoder { writeI32LE(this.reserve(4), v); return this; }
  i64(v: i64): BorshEncoder { writeI64LE(this.reserve(8), v); return this; }
  i128(v: i128): BorshEncoder { writeI128LE(this.reserve(16), v); return this; }

  // ── bool ────────────────────────────────────────────────────────────
  bool(v: bool): BorshEncoder { writeU8LE(this.reserve(1), v ? 1 : 0); return this; }

  // ── fixed 32-byte values: address / hash32 (raw, no prefix) ─────────
  private fixed32(b: StaticArray<u8>): BorshEncoder {
    assert(b.length == BYTES32_LEN, "borsh: address/hash32 must be 32 bytes");
    memory.copy(this.reserve(BYTES32_LEN), changetype<usize>(b), BYTES32_LEN);
    return this;
  }
  address(a: Address): BorshEncoder { return this.fixed32(a); }
  hash32(h: Bytes32): BorshEncoder { return this.fixed32(h); }

  // ── length-prefixed byte string: u32_le(len) ‖ raw bytes ────────────
  bytes(b: StaticArray<u8>): BorshEncoder {
    this.u32(<u32>b.length);
    if (b.length > 0) memory.copy(this.reserve(b.length), changetype<usize>(b), b.length);
    return this;
  }

  // ── UTF-8 string: u32_le(byte_len) ‖ UTF-8 bytes (no NUL) ────────────
  string(s: string): BorshEncoder {
    const utf8 = String.UTF8.encode(s); // ArrayBuffer, not NUL-terminated
    const n = utf8.byteLength;
    this.u32(<u32>n);
    if (n > 0) memory.copy(this.reserve(n), changetype<usize>(utf8), n);
    return this;
  }

  // ── vec<u64>: u32_le(len) ‖ each element as 8-byte LE ────────────────
  // The representative vec for this slice; generic `vec<T>` arrives with
  // codegen (Phases 2-4). A borsh `Vec<u8>` is identical to `bytes()`.
  vecU64(v: u64[]): BorshEncoder {
    this.u32(<u32>v.length);
    for (let i = 0, n = v.length; i < n; i++) this.u64(unchecked(v[i]));
    return this;
  }

  // Number of bytes encoded so far.
  // @ts-ignore: decorator
  @inline get length(): i32 { return this.len; }

  // Copy the accumulated bytes into a fresh, exactly-sized StaticArray.
  toBytes(): StaticArray<u8> {
    const out = new StaticArray<u8>(this.len);
    if (this.len > 0) memory.copy(changetype<usize>(out), changetype<usize>(this.data), this.len);
    return out;
  }
}

// ─────────────────────────────────────────────────────────────────────
// Decoder
// ─────────────────────────────────────────────────────────────────────
//
// A forward cursor over a StaticArray<u8>. Stores the ARRAY REFERENCE
// (not a raw pointer): a bare `usize` is invisible to the GC, so caching
// only the pointer could let the buffer be collected mid-decode. Every
// read bounds-checks against the input length and traps on underrun.

export class BorshDecoder {
  private data: StaticArray<u8>;
  private pos: i32;

  constructor(data: StaticArray<u8>) {
    this.data = data;
    this.pos = 0;
  }

  // Claim `n` bytes, return a pointer to them, advance the cursor.
  private take(n: i32): usize {
    assert(this.pos + n <= this.data.length, "borsh: unexpected end of input");
    const p = changetype<usize>(this.data) + <usize>this.pos;
    this.pos += n;
    return p;
  }

  // ── fixed-width integers ────────────────────────────────────────────
  u8(): u8 { return readU8LE(this.take(1)); }
  u16(): u16 { return readU16LE(this.take(2)); }
  u32(): u32 { return readU32LE(this.take(4)); }
  u64(): u64 { return readU64LE(this.take(8)); }
  u128(): u128 { return readU128LE(this.take(16)); }

  i8(): i8 { return readI8LE(this.take(1)); }
  i16(): i16 { return readI16LE(this.take(2)); }
  i32(): i32 { return readI32LE(this.take(4)); }
  i64(): i64 { return readI64LE(this.take(8)); }
  i128(): i128 { return readI128LE(this.take(16)); }

  // ── bool ────────────────────────────────────────────────────────────
  bool(): bool { return readU8LE(this.take(1)) != 0; }

  // ── fixed 32-byte values ────────────────────────────────────────────
  address(): Address { return bytes32FromPtr(this.take(BYTES32_LEN)); }
  hash32(): Bytes32 { return bytes32FromPtr(this.take(BYTES32_LEN)); }

  // ── length-prefixed byte string ─────────────────────────────────────
  bytes(): StaticArray<u8> {
    const n = <i32>this.u32();
    const out = new StaticArray<u8>(n);
    if (n > 0) memory.copy(changetype<usize>(out), this.take(n), n);
    return out;
  }

  // ── UTF-8 string ────────────────────────────────────────────────────
  string(): string {
    const n = <i32>this.u32();
    const p = this.take(n);
    return String.UTF8.decodeUnsafe(p, n);
  }

  // ── vec<u64> ────────────────────────────────────────────────────────
  vecU64(): u64[] {
    const n = <i32>this.u32();
    const out = new Array<u64>(n);
    for (let i = 0; i < n; i++) unchecked(out[i] = this.u64());
    return out;
  }

  // Bytes not yet consumed. A well-formed decode of a complete value
  // ends with `remaining() == 0`.
  // @ts-ignore: decorator
  @inline get remaining(): i32 { return this.data.length - this.pos; }
}
